### DISCUSSION TITLE
Deployment outputs as a string?

### DISCUSSION THREAD

**Discussion by ghost** [vercel/vercel#5939] [Help]
Hi,

I previously deployed the same apollo server api to vercel in the past, but my deployment is now showing up as a string on vercel.

https://waterfall-server-qcprm63g4-waterfall.vercel.app

- The current version api works with `vercel dev` locally
- Local webpack build works with `node public/index.js` locally
- No errors show up on vercel logs during deploy

I've tried to deploy with and without vercel.json and using `tsc` and `webpack` builds with the same result. 

Please let me know if you have any ideas how I can proceed. Thanks!

**Comment by ghost** [comment]
My guess, it's not recognizing that it's a node project and deploying static. 

When `api` folder is included, it does build routes. No error in logs. Ill post resulting deploy...

**Comment by ghost** [comment]
Webpack config:
```
const nodeExternals = require('webpack-node-externals');
const path = require('path');

module.exports = {
  entry: './api/index.ts',
  target: 'node',
  externals: [nodeExternals()],
  mode: 'production',
  module: {
    rules: [
      {
        test: /\.tsx?$/,
        use: 'awesome-typescript-loader',
        exclude: /node_modules/,
      },
    ],
  },
  resolve: {
    extensions: ['.ts', '.js'],
  },
  output: {
    filename: 'index.js',
    path: path.resolve(__dirname, 'public'),
  },
};
```

**Comment by ghost** [comment]
<img width="800" alt="Screen Shot 2021-03-06 at 10 49 00 AM" src="https://user-images.githubusercontent.com/1240213/110212514-80e66b80-7e50-11eb-9450-89862f0ffd45.png">

**Comment by ghost** [comment]
After vercel builds all the routes (api folder included), the app still deploys as a string: https://waterfall-server-hnjkweb8y-waterfall.vercel.app/

Scratching head...

**Comment by ghost** [comment]
Logs: https://vercel.com/waterfall/waterfall-server/4oqCaVCosHCcrABTfnvKy78wtZ4T

**Comment by toy-crane** [comment]
i stucked in same problem. did you solve it??

**Comment by ghost** [comment]
In the past, Vercel would automagically recognize the deploy as a node server and build on their servers. I used that method for a while, and was not optimized:

Here's the optimization:
1. Building your server locally decreases deploy time to under 10 seconds.
2. Then, need to manually point to the build and instruct Vercel about the API endpoint.

`webpack.config.js`
```
module.exports = {
  entry: './api/index.ts',
  target: 'node',
  externals: [nodeExternals()],
  mode: 'production',
  module: {
    rules: [
      {
        test: /\.tsx?$/,
        use: 'awesome-typescript-loader',
        exclude: /node_modules/,
      },
    ],
  },
  resolve: {
    extensions: ['.ts', '.js'],
  },
  output: {
    filename: 'index.js',
    path: path.resolve(__dirname, 'public'),
  },
};
```

`vercel.json`
```
{
  "version": 2,
  "builds": [{ "src": "public/index.js", "use": "@vercel/node" }],
  "routes": [{ "src": "/", "dest": "public/index.js" }]
}
```

Then, you can deploy with a script:
```
"build": "npx webpack",
"deploy": "yarn run build && vercel --prod"
```
